lib/flow_builder.dart

```diff
@@ -92,21 +92,21 @@ class _FlowBuilderState<T> extends State<FlowBuilder<T>> {
   @override
   void didUpdateWidget(FlowBuilder<T> oldWidget) {
     super.didUpdateWidget(oldWidget);
+
+    final _shouldDispose = oldWidget.controller == null;
+
+    _removeListeners(dispose: _shouldDispose);
+
     if (oldWidget.state != widget.state) {
-      _removeListeners(dispose: oldWidget.controller == null);
       _controller = _initController(widget.state);
-      _pages = widget.onGeneratePages(_state, List.of(_pages));
-      _history
-        ..clear()
-        ..add(_state);
     } else if (oldWidget.controller != widget.controller) {
-      _removeListeners(dispose: oldWidget.controller == null);
       _controller = widget.controller ?? _initController(_controller.state);
-      _pages = widget.onGeneratePages(_state, List.of(_pages));
-      _history
-        ..clear()
-        ..add(_state);
     }
+
+    _pages = widget.onGeneratePages(_state, List.of(_pages));
```

Comment:
Hmm I don't think we need to update the pages and history when the state and controller haven't changed. Thoughts?

Reply:
I think you're right. I didn't pay attention to this.

Then I understand why you did not reused this code, it will be hard to maintain and since you just used the same code twice, I think we're good!

Reply:
No worries! I really appreciate you taking the time to try to improve the code 🙏